run/oncemap.go

```diff
@@ -0,0 +1,43 @@
+// Copyright 2023 Terramate GmbH
+// SPDX-License-Identifier: MPL-2.0
+
+package run
+
+import "sync"
+
+// OnceMap contains map of key/value pairs that is safe for concurrent initialization.
+type OnceMap[K ~string, V any] struct {
+	mtx  sync.RWMutex
+	data map[K]V
+}
+
+// NewOnceMap returns a new empty OnceMap.
+func NewOnceMap[K ~string, V any]() *OnceMap[K, V] {
+	return &OnceMap[K, V]{data: make(map[K]V)}
+}
+
+// GetOrInit obtains a value for the given key if the value exists in the map, or
+// initializes it with the given init function.
+// This function is safe to be called concurrently.
+func (m *OnceMap[K, V]) GetOrInit(k K, init func() (V, error)) (V, error) {
+	m.mtx.RLock()
+	v, found := m.data[k]
+	m.mtx.RUnlock()
+
+	if !found {
```

Comment:
I'd say to use early return here to ease the readibility:
```suggestion
	if found {
	    return v, nil
        }
```

Reply:
Done in [9e11aeb](https://github.com/terramate-io/terramate/pull/2024/commits/9e11aeb6457a550b0e0d30ffa68c2fdb5c03d61d)